docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,798 @@
+
+```
+te/<name-space-1>/<name-space-2>/<name-space-3>/<name-space-4>/<data-type>/<data-instance-type>
+```
+
+Having 4 levels for entity identification provides enough flexibility
+to better namespace the devices and applications running on it minimizing the risk of `id` conflicts.
+Thin-edge doesn't enforce/dictate how a user uses these namespace levels.
+But there are conventions that can be followed to cover a wide range of use-cases, as follows:
+
+* `te/<device-namespace>/<device-id>/<service-namespace>/<service-id>/...`
+* `te/<device-namespace>/<device-id>/<device-local-namespace>/<device-local-component-id>/...`
+* `te/<entity-namespace>/<entity-id>/<component-namespace>/<component-id>/...`
+
+If your use-case doesn't need so many levels to identify an entity,
+the unnecessary levels can be skipped with the `_` (underscore) character as follows:
+
+```
+te/<entity-id>/_/_/_/...
+```
+
+or 
+
+```
+te/_/_/_/<entity-id>/...
+```
+
+Even for the data type levels, a user is free to define those as they wish.
+But thin-edge has some pre-defined `<data-type>` subtopics for well-known types like:
+`measurements`, `alarms`, `events` and `commands`, on which it enforces some constraints.
+
+**Subtopics for predefined telemetry data**
+
+|Type|Example|
+|------|-------|
+|Measurements|`<entity_id_prefix>/mes/[/<measurement-type>]`|
+|Events|`<entity_id_prefix>/evn/<event-type>`|
+|Alarms|`<entity_id_prefix>/alr/<alarm-type>`|
+
+**Subtopics for predefined commands**
+
+|Type|Example|
+|------|-------|
+|Software List|`<entity_id_prefix>/cmd/software_list`|
+|Software Update|`<entity_id_prefix>/cmd/software_update`|
+|Configuration Snapshot|`<entity_id_prefix>/cmd/config_snapshot`|
+|Configuration Update|`<entity_id_prefix>/cmd/config_update`|
+|Firmware Update|`<entity_id_prefix>/cmd/firmware_update`|
+|Restart[^1]|`<entity_id_prefix>/cmd/restart`|
+
+[^1]: Restart would mean a device restart or a service restart based on the target entity
+
+### Entity registration
+
+Since thin-edge doesn't enforce what each device identification level means,
+an explicit registration is required to register every entity that is going to send data or receive commands.
+For example, before a measurement can be sent from a service named `collectd` from the device `Rpi1001`,
+the entity named `Rpi1001` must be registered as a `device`, and `collectd` must be registered as a `service`.
+
+An entity can be registered with thin-edge by publishing a retained message to the entity identification topic prefix
+with the entity type and other metadata that defines that entity.
+To model the example mentioned above, an entity identification topic scheme like the following can be used:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+The `Rpi1001` device must be registered as a `device` by publishing the following retained message:
+
+```sh te2mqtt
+tedge mqtt pub -r tedge/_/Rpi1001 '{
```

Comment:
```suggestion
tedge mqtt pub -r tedge/_/Rpi1001/meta '{
```
Although not required, I'm really tempted to add this `/meta` subtopic level for entity registrations as well, to keep it consistent with data type metadata/

Reply:
Having `/meta` is not needed here and it just further complicates the topic structure. Just subscribing to `tedge/+/Rpi1001` is enough to get information about it, as that topic name does not conflict with anything else.

Reply:
The reason why the telemetry data has the `/meta` segment is that the parent topic is used for the value, where for the entities/components, these do not have inherent values so we can use it to describe the meta information about the device.

Reply:
I'm aligned with Reuben: such a `meta` suffix is useless and even annoying as introducing a special case with a reserved word.

Reply:
This idea has not been adopted.